src/mds/MDSKafka.cc

```diff
@@ -0,0 +1,526 @@
+  for (auto &[hash_key, endpoint] : endpoints) {
+    if (endpoint->has_topic(topic_name)) {
+      kafka_from = endpoint;
+      break;
+    }
+  }
+  if (kafka_from) {
+    kafka_from->remove_topic(topic_name);
+    if (kafka_from->topics.size() == 0) {
+      endpoints.erase(kafka_from->connection.hash_key);
+    }
+  }
+}
+
+void MDSKafkaManager::add_topic(const std::string &topic_name,
+                                const connection_t &connection) {
+  std::unique_lock<std::shared_mutex> lock(endpoint_mutex);
+  std::shared_ptr<MDSKafka> kafka_from, kafka_to;
+  for (auto &[hash_key, endpoint] : endpoints) {
+    if (endpoint->has_topic(topic_name)) {
+      kafka_from = endpoint;
+      break;
+    }
+  }
+  auto it = endpoints.find(connection.hash_key);
+  if (it != endpoints.end()) {
+    kafka_to = it->second;
+  }
+  if (kafka_from && kafka_from == kafka_to) {
+    ldout(cct, 1) << "Kafka topic creation error: kafka topic with name="
+                  << topic_name << " exists with same configuration" << dendl;
+    return;
+  }
+  bool created = false;
+  if (!kafka_to) {
+    if (endpoints.size() >= MAX_CONNECTIONS_DEFAULT) {
+      ldout(cct, 1) << "Kafka connect: max connections exceeded" << dendl;
+      return;
+    }
+    kafka_to = MDSKafka::create(cct, connection);
+    if (!kafka_to) {
+      return;
+    }
+    created = true;
+  }
+  std::shared_ptr<MDSKafkaTopic> topic =
+      MDSKafkaTopic::create(cct, topic_name, kafka_to);
+  if (!topic) {
+    return;
+  }
+  kafka_to->add_topic(topic_name, topic);
+  if (created) {
+    endpoints[connection.hash_key] = kafka_to;
+  }
+  if (kafka_from) {
+    kafka_from->remove_topic(topic_name);
+    if (kafka_from->topics.size() == 0) {
+      endpoints.erase(kafka_from->connection.hash_key);
+    }
+  }
+}
+
+int MDSKafkaManager::send(
+    const std::shared_ptr<MDSNotificationMessage> &message) {
+  std::unique_lock<std::mutex> lock(queue_mutex);
+  if (message_queue.size() >= MAX_QUEUE_DEFAULT) {
+    return -EBUSY;
+  }
+  message_queue.push(message);
+  return 0;
+}
+
+uint64_t MDSKafkaManager::publish(
+    const std::shared_ptr<MDSNotificationMessage> &message) {
+  std::shared_lock<std::shared_mutex> lock(endpoint_mutex);
+  uint64_t reply_count = 0;
+  for (auto &[key, endpoint] : endpoints) {
+    reply_count += endpoint->publish_internal(message);
+  }
+  return reply_count;
+}
+
+uint64_t MDSKafkaManager::polling() {
+  std::shared_lock<std::shared_mutex> lock(endpoint_mutex);
+  uint64_t reply_count = 0;
+  for (auto &[key, endpoint] : endpoints) {
+    reply_count += endpoint->poll();
+  }
+  return reply_count;
+}
+
+void MDSKafkaManager::run() {
+  while (!stopped) {
+    bool f = false;
+    while (true) {
+      std::unique_lock<std::mutex> lock(queue_mutex);
+      if (message_queue.empty()) {
+        break;
+      }
+      std::shared_ptr<MDSNotificationMessage> message = message_queue.front();
+      message_queue.pop();
+      f = true;
```

Comment:
Sorry which variable is unused here? `f` is used to indicate that at least one message is sent, if it is `false` and no replies are there then thread will sleep for some time to avoid busy loop.